src/Cli/dotnet/ReleasePropertyProjectLocator.cs

```diff
@@ -0,0 +1,160 @@
+    {
+        /// <param name="slnProjectPropertytoCheck">A property to enforce if we are looking into SLN files. If projects disagree on the property, throws exception.</param>
+        /// <returns>A project instance that will be targeted to publish/pack, etc. null if one does not exist.</returns>
+        public override ProjectInstance GetTargetedProject(IEnumerable<string> slnOrProjectArgs, string slnProjectPropertytoCheck = "")
+        {
+            string potentialProject = "";
+
+            foreach (string arg in slnOrProjectArgs.Append(Directory.GetCurrentDirectory()))
+            {
+                if (IsValidProjectFilePath(arg))
+                {
+                    return TryGetProjectInstance(arg);
+                }
+                else if (Directory.Exists(arg)) // We should get here if the user did not provide a .proj or a .sln
+                {
+                    try
+                    {
+                        return TryGetProjectInstance(MsbuildProject.GetProjectFileFromDirectory(arg).FullName);
+                    }
+                    catch (GracefulException)
+                    {
+                        // Fall back to looking for a solution if multiple project files are found.
+                        string potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
+
+                        if (!string.IsNullOrEmpty(potentialSln))
+                            return GetSlnProject(potentialSln, slnProjectPropertytoCheck);
+                    } // If nothing can be found: that's caught by MSBuild XMake::ProcessProjectSwitch -- don't change the behavior by failing here. 
+                }
+            }
+
+            return string.IsNullOrEmpty(potentialProject) ? null : TryGetProjectInstance(potentialProject);
+        }
+
+        /// <returns>The top-level project (first if multiple exist) in a SLN. Returns null if no top level project. Throws exception if two top level projects disagree
+        /// in the configuration property to check.</returns>
+        public override ProjectInstance GetSlnProject(string slnPath, string slnProjectConfigPropertytoCheck = "")
+        {
+            // This has a performance overhead so don't do this unless opted in.
+            if (
+                // PublishRelease v
+                (slnProjectConfigPropertytoCheck == PublishCommandParser.customDefaultConfigurationProperty && Environment.GetEnvironmentVariable(PublishCommandParser.enablePublishReleaseSolutionEnvironmentVar) == null)
+                || // PackRelease v
+                (slnProjectConfigPropertytoCheck == PackCommandParser.customDefaultConfigurationProperty && Environment.GetEnvironmentVariable(PackCommandParser.enablePackReleaseSolutionEnvironmentVar) == null)
+                )
+                return null; // The user will be warned if they do not have this set and try this scenario with one of the properties set.
```

Comment:
We're only ever expecting `slnProjectConfigPropertytoCheck` to equal `PackRelease` or `PublishRelease`.  It would be good to write this code so that if it gets any other values, it errors out so that we fix the code sending unexpected values, or to change it so it's not possible to pass an invalid value.

I would suggest checking for the environment variables in the `PublishCommand` and `PackCommand` classes, and then based on that passing a `bool` parameter through (or setting a property on the locator, or something) which controls whether it should look into solution files.

Reply:
Done, I added a class member and added an assert